boot: Add MCUBOOT_HW_KEY support for image encryption

DineshDK03 opened the pull request:
Currently encryption supports only private key embed in mcuboot itself. To support MCUBOOT_HW_KEY for image encryption boot_retrieve_private_key() hook is added.

This hook helps retrieving private key from trusted sources like OTP, TPM.

d3zd3z commented on boot/bootutil/src/encrypted.c:
```diff
@@ -513,6 +513,13 @@ boot_enc_decrypt(const uint8_t *buf, uint8_t *enckey)
+#if defined(MCUBOOT_HW_KEY)
+    rc = boot_retrieve_private_key(&bootutil_enc_key.key, &bootutil_enc_key.len);
+    if (rc) {
+	return rc;
```
Make sure your editor is configured to insert spaces instead of tabs when working on mcuboot source. In this case, this should be 8 spaces rather than a tab.

DineshDK03 replied:
done

d3zd3z reviewed (changes requested):
I like the idea of this change. I'd like to see if we can do this without introducing more ifdefs throughout the code. Perhaps we make the main code so that it always calls this function, and then, depending on the feature (ifdefs around whole functions are better than ones sprinkled through the code), we could just return a copy of the compiled in one. This could be conditionalized out if a given developer wishes to provide their own implementation.

DineshDK03 commented:
@d3zd3z  please revisit.

DineshDK03 commented:
@d3zd3z Ping

mingulov commented on boot/bootutil/src/encrypted.c:
```diff
@@ -438,8 +438,14 @@ boot_enc_decrypt(const uint8_t *buf, uint8_t *enckey)
     uint8_t counter[BOOTUTIL_CRYPTO_AES_CTR_BLOCK_SIZE];
     uint16_t len;
 #endif
+    const struct bootutil_key bootutil_enc_key = {0};
```
const struct means that this local `bootutil_enc_key` will remain {0} after a call to boot_retrieve_private_key

mingulov commented on boot/cypress/MCUBootApp/keys.c:
```diff
@@ -167,3 +167,12 @@ const struct bootutil_key bootutil_enc_key = {
+#if !defined(MCUBOOT_HW_KEY)
+int boot_retrieve_private_key(const struct bootutil_key *private_key)
+{
+        private_key = &bootutil_enc_key;
```
Ok as I see - seems this pull request had been broken in the 2nd version, an original version had no such problem:
boot_retrieve_private_key(&bootutil_enc_key.key, &bootutil_enc_key.len);

DineshDK03 replied:
Thanks for the review and sorry, was on a long vacation, I will go through this latest version and update respectively

mingulov commented on boot/bootutil/src/encrypted.c:
```diff
@@ -67,7 +67,7 @@ static int bootutil_constant_time_compare(const uint8_t *a, const uint8_t *b, si
 #if defined(MCUBOOT_ENCRYPT_KW)
 static int
-key_unwrap(const uint8_t *wrapped, uint8_t *enckey)
+key_unwrap(const uint8_t *wrapped, uint8_t *enckey,struct bootutil_enc_key)
```
I think it should be "struct bootutil_key"

DineshDK03 replied:
sorry, yes it should be `struct bootutil_key bootutil_enc_key`. thanks for pointing it out. Since Zephyr doesn't support KW_ENCRYPT I haven't tested that part.

jimmyzumthurm commented:
Any update on this ? We have similar solution currently in our software by patching MCUboot, but would highly appreciate if that would be officially available

DineshDK03 commented:
> Any update on this ? We have similar solution currently in our software by patching MCUboot, but would highly appreciate if that would be officially available

@jimmyzumthurm. I have already addressed the comments from @mingulov and @d3zd3z  and waiting for their acceptance to merge this PR.

mingulov commented:
> > Any update on this ? We have similar solution currently in our software by patching MCUboot, but would highly appreciate if that would be officially available
> 
> @jimmyzumthurm. I have already addressed the comments from @mingulov and @d3zd3z and waiting for their acceptance to merge this PR.

I am not an official approver / maintainer for MCUboot, so unfortunately my acceptance would not help anyhow.

DineshDK03 commented:
@d3zd3z  @de-nordic @nordicjm please have a look.

DineshDK03 commented:
@d3zd3z ping

nordicjm commented:
@sigvartmh

DineshDK03 commented:
@d3zd3z @sigvartmh @nordicjm @de-nordic please have a look, if all is good, kindly merge this PR.

d3zd3z commented on boot/bootutil/src/encrypted.c:
```diff
@@ -67,7 +67,7 @@ static int bootutil_constant_time_compare(const uint8_t *a, const uint8_t *b, si
 #if defined(MCUBOOT_ENCRYPT_KW)
 static int
-key_unwrap(const uint8_t *wrapped, uint8_t *enckey)
+key_unwrap(const uint8_t *wrapped, uint8_t *enckey,struct bootutil_key bootutil_enc_key)
```
There should be a space after the comma.

DineshDK03 replied:
fixed

d3zd3z reviewed (changes requested):
Other than the whitespace issue, this looks good now.

DineshDK03 commented:
@d3zd3z Thanks for the review and fixed all the whitespace issues now.

DineshDK03 commented:
@d3zd3z pushed changes to fix the CI. please approve to run the workflow jobs.

d3zd3z commented on boot/mbed/app_enc_keys.c:
```diff
@@ -69,3 +69,12 @@ const struct bootutil_key bootutil_enc_key = {
+#if !defined(MCUBOOT_HW_KEY) && defined(MCUBOOT_ENC_IMAGES)
+int boot_retrieve_private_key(struct bootutil_key *private_key)
+{
+    private_key = (struct bootutil_key *)&bootutil_enc_key;
```
This line, and the ones below, doesn't actually do anything. It it just copying the address of the key to a local variable. I suspect what you actually want is for the parameter to be `struct bootutil_key **private_key`.

DineshDK03 replied:
fixed it now. please trigger CI

DineshDK03 replied:
@d3zd3z CI passed. I hope we can merge this now.

DineshDK03 commented:
@d3zd3z please have a look, all the comments addressed and waiting to be merged.

davidvincze commented on boot/bootutil/include/bootutil/enc_key.h:
```diff
@@ -45,7 +46,17 @@ struct enc_key_data {
+ * @return                   0 on success; nonzero on failure.
+ *
+ */
+int boot_retrieve_private_key(struct bootutil_key **private_key);
```
Would you mind renaming this function to `boot_enc_retrieve_priv_key` or similar?
It's for avoiding ambiguity - one might think that it's retrieving the private key used for image signing (there's already a function called `boot_retrieve_public_key_hash`).

davidvincze replied:
The renaming is still missing in one file. I've added a comment to it.

davidvincze commented on boot/cypress/MCUBootApp/keys.c:
```diff
@@ -167,3 +167,12 @@ const struct bootutil_key bootutil_enc_key = {
+
+    return 0;
+}
+#endif /* !MCUBOOT_HW_KEY */
```
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

davidvincze commented on boot/mbed/app_enc_keys.c:
```diff
@@ -69,3 +69,12 @@ const struct bootutil_key bootutil_enc_key = {
+
+    return 0;
+}
+#endif /* !MCUBOOT_HW_KEY */
```
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

davidvincze commented on boot/zephyr/keys.c:
```diff
@@ -86,3 +86,12 @@ const struct bootutil_key bootutil_enc_key = {
+
+    return 0;
+}
+#endif /* !MCUBOOT_HW_KEY */
```
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

davidvincze commented on ci/mynewt_keys/enc_kw/src/keys.c:
```diff
@@ -28,3 +28,12 @@ const struct bootutil_key bootutil_enc_key = {
+
+    return 0;
+}
+#endif /* !MCUBOOT_HW_KEY */
```
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

davidvincze commented on ci/mynewt_keys/enc_rsa/src/keys.c:
```diff
@@ -126,3 +126,12 @@ const struct bootutil_key bootutil_enc_key = {
+
+    return 0;
+}
+#endif /* !MCUBOOT_HW_KEY */
```
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

davidvincze commented on sim/mcuboot-sys/csupport/keys.c:
```diff
@@ -328,3 +328,12 @@ const struct bootutil_key bootutil_enc_key = {
+
+    return 0;
+}
+#endif /* !MCUBOOT_HW_KEY */
```
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

davidvincze commented:
@DineshDK03 please add a top commit that adds a release note snippet for this feature:
https://github.com/mcu-tools/mcuboot/blob/main/docs/SubmittingPatches.md#release-notes
Thank you!

davidvincze commented on sim/mcuboot-sys/csupport/keys.c:
```diff
@@ -328,3 +328,12 @@ const struct bootutil_key bootutil_enc_key = {
 #endif
+
+#if !defined(MCUBOOT_HW_KEY) && defined(MCUBOOT_ENC_IMAGES)
+int boot_retrieve_private_key(struct bootutil_key **private_key)
```
```suggestion
int boot_enc_retrieve_private_key(struct bootutil_key **private_key)
```

DineshDK03 replied:
missed it, updated now, thanks.

davidvincze commented:
@d3zd3z any other comments on this PR or is it good to go?
